src/workerd/api/streams/internal.c++

```diff
@@ -96,23 +94,24 @@ private:
   kj::Vector<kj::Array<kj::byte>> parts;
   uint64_t runningTotal = 0;
 
-  kj::Promise<PartList> loop() {
-    auto bytes = kj::heapArray<kj::byte>(4096);
+  kj::Promise<PartList> readParts() {
+    auto buffer = kj::heapArray<kj::byte>(4096);
+
+    while (true) {
+      auto amount = co_await input.tryRead(buffer.begin(), 1, buffer.size());
 
-    return input.tryRead(bytes.begin(), 1, bytes.size())
-        .then([this, bytes = kj::mv(bytes)](size_t amount) mutable
-        -> kj::Promise<PartList> {
       if (amount == 0) {
-        return KJ_MAP(p, parts) { return p.asPtr(); };
+        co_return KJ_MAP(p, parts) { return p.asPtr(); };
       }
 
       runningTotal += amount;
       if (runningTotal >= limit) {
-        return JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
+        throw JSG_KJ_EXCEPTION(FAILED, TypeError, "Memory limit exceeded before EOF.");
       }
-      parts.add(bytes.slice(0, amount).attach(kj::mv(bytes)));
-      return loop();
-    });
+      auto part = kj::heapArray<kj::byte>(amount);
+      memcpy(part.begin(), buffer.begin(), amount);
```

Comment:
why the change from `bytes.slice` to `memcpy` here? I'd consider `slice` to be safer.

Reply:
Before my change it would allocate several 4K buffers and result would be slices pointing into these buffers. Now we have one buffer. That's why we need to copy its data somewhere.

Reply:
I managed to remove memcpy anyway. ptal.